### ISSUE TITLE
Network IO between kafka Nodes

### ISSUE THREAD

**Issue by hikrrish** [allegro/hermes#543]
@adamdubiel 
We are seeing a network data transfer of 20 GB per hour among 3 nodes of  kafka brokers running on AWS deployed in 3 different AZ. This is way beyond our actual data transmission in the system, which is around 200 MB per hour with message retention time set to 40 days. The actual data on each of kafka node is around 10 GB. Our kafka cluster is only utilized by hermes apps and are not flowing any data via any other channels. I believe kafka Producer configurations are managed by hermes only so we have tried using snappy as compression algorithm. do you see any tuning configuration in frontend and consumer to tune this issue?

Thanks

**Comment by adamdubiel** [comment]
Is there any traffic on Hermes? Hermes only sends traffic out that comes in. I think the traffic between Kafka might come from replication? Do you send data with ACK-ALL, so it makes some more requests to make sure dat is on all brokers?

Did you do tests without Hermes to see how vanilla Kafka behaves?

**Comment by hikrrish** [comment]
Hermes seems to be fine in terms of traffic, we are using leader ack only, currently kafka brokers are used by Hermes only and not other traffic is routed directly to kafka. Will test with replication factor =1 to see if that makes any difference.

**Comment by adamdubiel** [comment]
I can ask our Kafka ops about the traffic they see between hosts and compare it with hermes incoming traffic if it can be of any help to you.

**Comment by gamefundas** [comment]
It would be helpful Adam as we see that the Kafka nodes are syncing the entire logs continuously, regardless of any incoming messages. This is adding a lot of AWS interzone traffic cost. We enabled log compression to reduce some overhead but it would be nice to solve this. Didnt find any other configuration in Kafka that will control this behavior. 

Appreciate your help on this.

**Comment by adamdubiel** [comment]
Hi guys, i'm sorry but i never got the data from our clusters to send to you :/ I will continue on poking Kafka ops.

**Comment by gamefundas** [comment]
@adamdubiel sorry for bugging you again on this but we ran in to few issues on production with Hermes/Kafka wherein one of the consumer stopped reading from its associated Kafka partitions. Restarting consumer didn't work either and we had to eventually restart the entire cluster. We did try different things prior to starting fresh like restarting Zk, all Hermes consumers, all kafka nodes etc. but failed to recover. Consumer logs didn't indicate anything at all as we failed to understand what was preventing it to read those messages. Kafka manager reported no lag for the other consumer while this one remained with a lag of 2K+ odd messages.

It seems we haven't learnt,enough to manage the pubsub infrastructure. Ours is a limited setup with 3 Zk nodes, 3 Kafka nodes, 2 Hermes frontend, 2 Hermes Consumer, Hermes sharing the ZK instances with Kafka. We did enable ACK all, have 3 replicas and 3 partitions per broker. Not sure if we have configured this correctly as we were thinking the 3 partitions is not the right distribution for 2 consumers and were planning to increase partitions to 6.

Do you think we are missing something here? Also is there some kind of recommended cluster configuration, min no of nodes of each component or kafka/zk settings that your Ops team has which is well suited for running Hermes. Please let us know.

Thanks
Andy

**Comment by adamdubiel** [comment]
The setup is perfectly fine. I would start from switching to selective algorithm. But all in all - i see that you run into different kinds of problems and it's hard to help without knowing the context. Maybe we can get on video call just to talk, maybe there will be some things i can help with immediately?

**Comment by gamefundas** [comment]
Sorry had been away for a while and couldn't respond back to the thread. 
Hugely appreciate your offer to help. We did manage to sort some of the issues on our side in the past few weeks to get a stable prod environment.

Essentially in our AWS setup we changed cloud formation to use static IP's so that zookeeper and its state doesn't get confused with ephemeral ip's. We also upgraded to Kafka 0.10 which helped significantly. Few new issues emerged but @hikrrish  was able to resolve them with your teams help.

Thanks again for all the good work. We should be able to close this thread for now.
